stan/math/prim/fun/lb_constrain.hpp

```diff
@@ -44,29 +42,24 @@ inline return_type_t<T, L> lb_constrain(const T& x, const L& lb) {
  * reference with the log absolute Jacobian determinant of the
  * transform.
  *
- * If the lower bound is negative infinity, this function
- * reduces to <code>identity_constraint(x, lp)</code>.
- *
- * @tparam T type of scalar
+ * @tparam T Type of Matrix
  * @tparam L type of lower bound
  * @tparam S type of log probability
- * @param[in] x unconstrained scalar input
+ * @param[in] x unconstrained Matrix input
  * @param[in] lb lower bound on output
  * @param[in,out] lp reference to log probability to increment
  * @return lower-bound constrained value corresponding to inputs
  */
-template <typename T, typename L, typename S>
-inline return_type_t<T, L> lb_constrain(const T& x, const L& lb, S& lp) {
-  using std::exp;
-  if (lb == NEGATIVE_INFTY) {
-    return identity_constrain(x, lp);
-  }
-  lp += x;
-  return exp(x) + lb;
+template <typename T, typename L>
+inline auto lb_constrain(const T& x, const L& lb, return_type_t<T, L>& lp) {
+  const auto& x_ref = to_ref(x);
+  const auto& lb_ref = to_ref(lb);
+  check_finite("lb_constrain", "lb", value_of(lb_ref));
+  lp += sum(x_ref);
+  return eval(add(exp(x_ref), lb));
```

Comment:
Just remove all the eval's in these functions

Reply:
I took the evals out and got (deleting a bunch of output to make the picture clearer):

```
==4440==ERROR: AddressSanitizer: heap-use-after-free on address 0x603000ad0b90 at pc 0x000000992c36 bp 
...
 /home/bbales2/math-benchmarks/math/./test/unit/math/test_ad.hpp:86:22

0x603000ad0b90 is located 0 bytes inside of 32-byte region [0x603000ad0b90,0x603000ad0bb0)
freed by thread T0 here:
...
    #2  /home/bbales2/math-benchmarks/math/./stan/math/prim/fun/lub_constrain.hpp:106:1

previously allocated by thread T0 here:
...
    #5 /home/bbales2/math-benchmarks/math/./stan/math/prim/fun/to_ref.hpp:18:10
    #6  /home/bbales2/math-benchmarks/math/./stan/math/prim/fun/lub_constrain.hpp:101:22
```

That `to_ref` is the one that was creating the intermediate for `subtract`. It's possible any of the `to_ref` s could allocate memory here if they are expressions.